Implement runtime deps tracker

pmvald opened the pull request:
An initial implementation which captures what we need for both JIT and local compilations. This PR only changes deps tracker files which are NOT being used anywhere yet. So this change does not affect anything. In a follow up PR I'll migrate JIT compilation to use this deps tracker (instead of transitiveScopesFor tools and monkey patching) and hopefully it will break lots of tests cross g3 and worldwide which makes it look like I have done something ...

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

pmvald committed:
refactor(core): implement getNgModuleScope method of runtime dependency tracker

The implementation is more or less follows the pattern in render3/jit/module.ts#transitiveScopesFor helper. A few additional helper functions also added to jit utils.

pmvald committed:
refactor(core): implement getStandaloneComponentScope method of the runtime deps tracker

The logic mainly followed the `render3/jit/directive.ts#getStandaloneDefFunctions` helper.

pmvald committed:
refactor(core): implement clearScopeCacheFor for runtime deps tracker

This method mainly has application test beds where we want to apply overrides and re-compute the scope.

pmvald committed:
refactor(core): implement logic for getting component dependencies in the runtime deps tracker

This includes implementation of methods getComponentDependencies and registerNgModule.

In order to correlate ng-modules with their declarations it is required to use the method registerNgModule to regiater the ng-module. However, the actual correlation will happen lazily once getComponentDependencies method is called. This lazy behaviour also allows for forward refs to be resolved.

The method getComponentDependencies will be used in local compilation mode to compute the rendering component deps in runtime.

pmvald commented on packages/core/src/render3/deps_tracker/api.ts:
```diff
@@ -58,10 +58,14 @@ export interface DepsTrackerApi {
    * present in the component's template (This set might contain directives/components/pipes not
    * necessarily used in the component's template depending on the implementation).
    *
+   * Standalone components should specify `rawImports` as this information is not available from
```
For local compilation yes, the imports are stored in the class using runtime `ɵsetClassMetadata`, so we can ditch rawImports here. For JIT this method will be used in `getStandaloneDefFunctions` (https://github.com/angular/angular/blob/25153e94f60f7be055d472ab573f4e367f53cbe8/packages/core/src/render3/jit/directive.ts#L237) which passes rawImports as a param rather than stored inside the type.

I was also thinking if we store imports inside the component type, like the way we do for ngmodules, then probably things become more consistent. We did not discuss this during the last meeting so I thought I don't change things.

pmvald replied:
Done!

devversion commented on packages/core/src/render3/deps_tracker/deps_tracker.ts:
```diff
@@ -17,38 +22,293 @@ import {ComponentDependencies, DepsTrackerApi, NgModuleScope, StandaloneComponen
+      return this.ngModulesScopeCache.get(type)!;
+    }
+
+    const ans = this.computeNgModuleScope(type);
```
Nit

```suggestion
    const scope = this.computeNgModuleScope(type);
```

devversion commented on packages/core/src/render3/deps_tracker/deps_tracker.ts:
```diff
@@ -17,38 +22,293 @@ import {ComponentDependencies, DepsTrackerApi, NgModuleScope, StandaloneComponen
+      if (isNgModule(imported)) {
+        moduleToProcess = imported;
+      } else if (isModuleWithProviders(imported)) {
+        moduleToProcess = (type as any).ngModule;
```
should this be `imported.ngModule`?

devversion commented on packages/core/src/render3/deps_tracker/deps_tracker.ts:
```diff
@@ -17,38 +22,293 @@ import {ComponentDependencies, DepsTrackerApi, NgModuleScope, StandaloneComponen
+  /** Compute NgModule scope afresh. */
+  private computeNgModuleScope(type: NgModuleType<any>): NgModuleScope {
+    const def = getNgModuleDef(type, true);
+    const ans: NgModuleScope = {
```
```suggestion
    const scope: NgModuleScope = {
```

devversion commented on packages/core/src/render3/deps_tracker/deps_tracker.ts:
```diff
@@ -17,38 +22,293 @@ import {ComponentDependencies, DepsTrackerApi, NgModuleScope, StandaloneComponen
+          ans.compilation.pipes.add(imported);
+        } else {
+          // The standalone thing is neither a component nor a directive nor a pipe ... (what?) So
+          // we short circuit here.
```
do we want to log something here? or leave a TODO

devversion commented on packages/core/src/render3/deps_tracker/deps_tracker.ts:
```diff
@@ -17,38 +22,293 @@ import {ComponentDependencies, DepsTrackerApi, NgModuleScope, StandaloneComponen
+    }
+
+    // Analyzing declarations
+    if (!ans.compilation.isPoisoned) {
```
can we remove some of these conditions and return early if compilation is poisoned to avoid nesting?

pmvald replied:
Simplified some of the if logics.

In theory the only case we can return early is when both compilation and exported scopes are poisoned (otherwise there is still hope to compute the non-poisoned one ... )  On the other hand I see there is no way to have poisoned exported scope based on the current logic. So by removing check for exported scope I made some simplifications.

devversion commented on packages/core/src/render3/deps_tracker/deps_tracker.ts:
```diff
@@ -17,38 +22,293 @@ import {ComponentDependencies, DepsTrackerApi, NgModuleScope, StandaloneComponen
+            return ans;
+          }
+
+          if (!ans.compilation.isPoisoned) {
```
Will we ever get here since we `return` above? or can we remove this check?

pmvald replied:
Removed the check above

devversion reviewed (approved):
overall LGTM. this is pretty clean!

pmvald commented on packages/core/src/render3/deps_tracker/api.ts:
```diff
@@ -90,11 +94,16 @@ export interface DepsTrackerApi {
    */
-  getStandaloneComponentScope(type: ComponentType<any>, imports: Type<any>[]):
-      StandaloneComponentScope;
+  getStandaloneComponentScope(
```
Yes, it will be used in getStandaloneDefFunctions in JIT compiler.

pmvald commented on packages/core/src/render3/deps_tracker/deps_tracker.ts:
```diff
@@ -17,38 +22,293 @@ import {ComponentDependencies, DepsTrackerApi, NgModuleScope, StandaloneComponen
+    this.resolveNgModulesDecls();
+
+    const def = getComponentDef(type);
+    if (!def) {
```
Done!

pmvald commented on packages/core/src/render3/deps_tracker/deps_tracker.ts:
```diff
@@ -17,38 +22,293 @@ import {ComponentDependencies, DepsTrackerApi, NgModuleScope, StandaloneComponen
+      throw new Error(`Attempting to register a Type which is not NgModule as NgModule: ${type}`);
+    }
+
+    // Lazily process the NgModules later when need be.
```
Done!

devversion reviewed (approved):
LGTM overall, just some minor comments

devversion commented on packages/core/src/render3/deps_tracker/deps_tracker.ts:
```diff
@@ -17,38 +22,283 @@ import {ComponentDependencies, DepsTrackerApi, NgModuleScope, StandaloneComponen
+
+        // Based on the current logic there is no way to have poisoned exporter scope. So no need to
+        // check for it.
+        addSet(exportedScope.exported.directives, scope.exported.directives);
```
Optional: Do you think we can somehow reflect that lack of handling in the types of `.exported`?

pmvald commented on packages/core/test/render3/deps_tracker_spec.ts:
```diff
@@ -0,0 +1,1169 @@
+
+           const ans = depsTracker.getNgModuleScope(MainModule as NgModuleType);
+
+           expect(ans.compilation).toEqual({
```
After some thought I realized it is better to follow the AoT style to keep local compilation and AoT same. Did TGP and nothing was broken when JIT uses deps tracker. So safe to do this.

pmvald committed:
test(core): update runtime error list to include deps tracker error

The previous commits add a new runtime error code (RUNTIME_DEPS_INVALID_IMPORTED_TYPE) which needs to be added to the golden for the tests to pass.

jessicajaniuk reviewed (approved):
reviewed-for: public-api

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api

jessicajaniuk commented:
This PR was merged into the repository by commit 89c84bf2c1901bef4f2ec61baf200335ef9cd91d.
